plugins/droid/droid-codec.cpp

```diff
@@ -0,0 +1,725 @@
+        droid_media_codec_set_data_callbacks(m_codec, &cb, this);
+    }
+
+    if (!droid_media_codec_start (m_codec)) {
+        droid_media_codec_destroy (m_codec);
+        m_codec = nullptr;
+        LOGE("Failed to start the decoder");
+        return false;
+    }
+
+    configureOutput();
+
+    LOGD("Decoder started for " << m_metadata.parent.type);
+
+    return true;
+}
+
+// This is called when the user has not provided a release callback.
+void DroidVideoDecoder::dummyRelease(void *)
+{
+}
+
+bool DroidVideoDecoder::decode(const uint8_t *data,
+                               size_t size,
+                               uint64_t timestampUs,
+                               FrameType frameType,
+                               void (*release)(void *),
+                               void *releaseData)
+{
+    DroidMediaBufferCallbacks cb;
+    DroidMediaCodecData cdata;
+
+    LOGV("Decode: timestamp=" << timestampUs << " frameType" << frameType);
+
+    if (!m_codec && !createCodec()) {
+        LOGE("Cannot create decoder");
+        return false;
+    }
+
+    cdata.ts = timestampUs;
+    cdata.sync = frameType == KeyFrame;
+    cdata.data.size = size;
+    cdata.data.data = (void *)(data);
+
+    cb.data = releaseData;
+    cb.unref = release ? release : DroidVideoDecoder::dummyRelease;
+
+    // This blocks when the input queue is full
+    droid_media_codec_queue (m_codec, &cdata, &cb);
+
+    LOGV("Frame queued to decoder");
+
+    return true;
+}
+
+void DroidVideoDecoder::drain()
+{
+    LOGD("");
+    stop();
+}
+
+void DroidVideoDecoder::flush()
+{
+    LOGD("");
+    stop();
+}
+
+void DroidVideoDecoder::stop()
+{
+    if (m_codec) {
+        LOGD("");
+        droid_media_codec_drain(m_codec);
+        droid_media_codec_stop(m_codec);
+        droid_media_codec_destroy(m_codec);
+        m_codec = nullptr;
+    }
+}
+
+void DroidVideoDecoder::configureOutput()
+{
+    DroidMediaCodecMetaData md;
+    DroidMediaRect rect;
+
+    memset (&md, 0x0, sizeof (md));
+    memset (&rect, 0x0, sizeof (rect));
+
+    droid_media_codec_get_output_info (m_codec, &md, &rect);
+    LOGI("Configuring converter for"
+         << " stride:" << md.width
+         << " slice-height: " << md.height
+         << " top: " << rect.top << " left:" << rect.left
+         << " width: " << rect.right - rect.left
+         << " height: " << rect.bottom - rect.top
+         << " format: " << md.hal_format);
+    m_mapper.setFormat(&md, &rect);
+}
+
+void DroidVideoDecoder::error(string errorDescription)
+{
+    LOGE(errorDescription);
+
+    if (m_decoderListener) {
+        m_decoderListener->onDecoderError(errorDescription);
+    }
+}
+
+void DroidVideoDecoder::dataAvailable(DroidMediaCodecData *decoded)
+{
+    if (m_decoderListener && m_mapper.ready()) {
+        gecko::camera::YCbCrFrame frame = m_mapper.mapYcbCr(decoded);
+        m_decoderListener->onDecodedFrame(frame);
+        // FIXME: Do I need to store duration?
```

Comment:
Is the FIXME safe not to fix at this stage?

Reply:
Duration is stored on the gecko side of the decoder, so I dropped the FIXME.